### PR TITLE
Fix #293: Incorrect handling of 'finally'?

### LINKED ISSUE #293: Incorrect handling of 'finally'?
```
Not sure if this is a bug or a misunderstanding on my end.

    public abstract class Init
    {
        public void f()
        {
            String s;
            try {
                s = read();
            }
            catch (Exception e) {
                return;
            }
            finally {
                write("finished");
            }

            write(s);   // <-- error reported here
        }

        public abstract String read() throws Exception;
        public abstract void write(String s);
    }

I expected the nullness checker to report no errors.

# java -jar binary/checkers.jar -processor checkers.nullness.NullnessChecker 
Init.java
Init.java:16: error: incompatible types in argument.
        write(s);
              ^
  found   : @UnknownInitialization @Nullable String
  required: @Initialized @NonNull String
1 error

Removing the 'finally' block entirely makes the error go away.
```

Original issue reported on code.google.com by `kan...@cakoose.com` on 18 Dec 2013 at 1:10

### PR DESCRIPTION
**Pull request by bitterfox**
In this PR, I fix incorrect handling of `finally`.
#341 is also fully fixed by this PR.

The cause of the problem is that `finally` blocks are treated as same for a successful path and exceptional path though they have a completely different path for next of corresponding `finally` block.
The next path for successful `finally` block will be next successful path and another will be an exceptional path, throw the exception which is thrown before `finally` block.
Because `finally` block for both path are treated as same, the input stores are merged, and it's cause of this problem.

I fix it to clone the finally block and make those stores are not merged.

### CODE REVIEW

File: dataflow/src/org/checkerframework/dataflow/cfg/ControlFlowGraph.java [hunk 1]
```diff
@@ -61,14 +61,18 @@
      */
     protected final List<ReturnNode> returnNodes;
 
+    /** Map from AST {@link Tree}s to generated {@link Tree}s. */
+    protected IdentityHashMap<Tree, List<Tree>> generatedTreesLookupMap;
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/cfg/ControlFlowGraph.java]: Can this field be `final`? I think the other non-final fields in this class can also be changed.

File: dataflow/src/org/checkerframework/dataflow/cfg/ControlFlowGraph.java [hunk 2]
```diff
@@ -216,6 +221,11 @@ public UnderlyingAST getUnderlyingAST() {
         return new IdentityHashMap<>(treeLookup);
     }
 
+    /** @return the map to lookup generated {@link Tree}s from {@link Tree} */
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/cfg/ControlFlowGraph.java]: Make clear that this returns a copy.

File: framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java [hunk 3]
```diff
@@ -884,6 +887,21 @@ public Void visitEnhancedForLoop(EnhancedForLoopTree node, Void p) {
         return super.visitEnhancedForLoop(node, p);
     }
 
+    @Override
+    public Void visitTry(TryTree node, Void aVoid) {
+        Void ret = super.visitTry(node, aVoid);
+        JCTry tree = (JCTry) node;
+        JCBlock oldFinalizer = tree.finalizer;
+        if (oldFinalizer != null) {
+            for (Tree generatedFinalizer : atypeFactory.getGeneratedTrees(oldFinalizer)) {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java]: Add a short comment documenting what this is doing.

File: javacutil/src/org/checkerframework/javacutil/trees/FullyTreeCopier.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+import com.sun.tools.javac.tree.JCTree.JCTry;
+import com.sun.tools.javac.tree.JCTree.JCUnary;
+import com.sun.tools.javac.tree.JCTree.JCVariableDecl;
+import com.sun.tools.javac.tree.TreeCopier;
+import com.sun.tools.javac.tree.TreeMaker;
+
+/** A Utility class for coping AST {@link JCTree} fully, including type and symbol information. */
+class FullyTreeCopier extends TreeCopier<Void> {
+
+    private TreeMaker M;
```
Comment (wmdietl) [inline javacutil/src/org/checkerframework/javacutil/trees/FullyTreeCopier.java]: `M` is not the best name for this field. Why not `treeMaker` or `maker`?
Can the field be final?

File: javacutil/src/org/checkerframework/javacutil/trees/FullyTreeCopier.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+import com.sun.tools.javac.tree.JCTree.JCMethodDecl;
+import com.sun.tools.javac.tree.JCTree.JCMethodInvocation;
+import com.sun.tools.javac.tree.JCTree.JCNewClass;
+import com.sun.tools.javac.tree.JCTree.JCTry;
+import com.sun.tools.javac.tree.JCTree.JCUnary;
+import com.sun.tools.javac.tree.JCTree.JCVariableDecl;
+import com.sun.tools.javac.tree.TreeCopier;
+import com.sun.tools.javac.tree.TreeMaker;
+
+/** A Utility class for coping AST {@link JCTree} fully, including type and symbol information. */
```
Comment (wmdietl) [inline javacutil/src/org/checkerframework/javacutil/trees/FullyTreeCopier.java]: Explain why the class has package visibility and where it is used.

File: javacutil/src/org/checkerframework/javacutil/trees/FullyTreeCopier.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+class FullyTreeCopier extends TreeCopier<Void> {
+
+    private TreeMaker M;
+
+    FullyTreeCopier(TreeMaker M) {
```
Comment (wmdietl) [inline javacutil/src/org/checkerframework/javacutil/trees/FullyTreeCopier.java]: Call out default visibility.

File: javacutil/src/org/checkerframework/javacutil/trees/FullyTreeCopier.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    public JCTree visitMethodInvocation(MethodInvocationTree node, Void p) {
+        JCMethodInvocation t = (JCMethodInvocation) super.visitMethodInvocation(node, p);
+        t.varargsElement = ((JCMethodInvocation) node).varargsElement;
+        return t;
+    }
+
+    public JCTree visitNewClass(NewClassTree node, Void p) {
+        JCNewClass t = (JCNewClass) node;
+        JCExpression encl = (JCExpression) super.visitNewClass(node, p);
+        t.constructor = ((JCNewClass) node).constructor;
```
Comment (wmdietl) [inline javacutil/src/org/checkerframework/javacutil/trees/FullyTreeCopier.java]: Isn't this a no-op? `t` is an alias for `node` and here you are casting `node` again.
Am I missing what all these assignments do?

File: dataflow/src/org/checkerframework/dataflow/cfg/CFGBuilder.java [hunk 8]
```diff
@@ -876,9 +879,56 @@ public static ControlFlowGraph process(ControlFlowGraph cfg) {
                 }
             }
 
+            // remove generated trees when corresponding nodes are unreachable
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/cfg/CFGBuilder.java]: Can you add a few more explanations of the steps below? I don't see where anything gets removed, you're building up the map of generatedTrees.

File: dataflow/src/org/checkerframework/dataflow/cfg/CFGBuilder.java [hunk 9]
```diff
@@ -876,9 +879,56 @@ public static ControlFlowGraph process(ControlFlowGraph cfg) {
...
+            cfg.generatedTreesLookupMap = generatedTreesLookupMap;
+
             return cfg;
         }
 
+        private static class ContainsAnyScanner extends TreeScanner<Void, Void> {
+            private boolean contains = false;
+            private final Set<Tree> trees;
+
+            ContainsAnyScanner(Set<Tree> trees) {
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/cfg/CFGBuilder.java]: Call out default visibility or use something more restrictive.

File: dataflow/src/org/checkerframework/dataflow/cfg/CFGBuilder.java [hunk 10]
```diff
@@ -876,9 +879,56 @@ public static ControlFlowGraph process(ControlFlowGraph cfg) {
...
+                        generatedTrees.add(root);
+                    }
+                }
+            }
+            cfg.generatedTreesLookupMap = generatedTreesLookupMap;
+
             return cfg;
         }
 
+        private static class ContainsAnyScanner extends TreeScanner<Void, Void> {
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/cfg/CFGBuilder.java]: Add a short description of what the class is for.

File: dataflow/src/org/checkerframework/dataflow/cfg/CFGBuilder.java [hunk 11]
```diff
@@ -1423,6 +1477,8 @@ public CFGTranslationPhaseOne() {}
          */
         private List<ReturnNode> returnNodes;
 
+        protected IdentityHashMap<Tree, List<Tree>> generatedTreesLookupMap;
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/cfg/CFGBuilder.java]: Some documentation.

File: dataflow/src/org/checkerframework/dataflow/cfg/CFGBuilder.java [hunk 12]
```diff
@@ -4095,20 +4170,58 @@ public Node visitTry(TryTree tree, Void p) {
...
-                scan(finallyBlock, p);
 
-                TypeMirror throwableType = elements.getTypeElement("java.lang.Throwable").asType();
-                extendWithNodeWithException(
-                        new MarkerNode(tree, "end of finally block", env.getTypeUtils()),
-                        throwableType);
+                boolean hasExceptionalPath = hasExceptionalPath(exceptionalFinallyLabel);
+
+                if (hasExceptionalPath) {
+                    addLabelForNextNode(exceptionalFinallyLabel);
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/cfg/CFGBuilder.java]: Add a brief description of what the two branches do.

File: dataflow/src/org/checkerframework/dataflow/cfg/CFGBuilder.java [hunk 13]
```diff
@@ -4095,20 +4170,58 @@ public Node visitTry(TryTree tree, Void p) {
...
+                    extendWithExtendedNode(new UnconditionalJump(doneLabel));
+                }
             }
 
             addLabelForNextNode(doneLabel);
 
             return null;
         }
 
+        private boolean hasExceptionalPath(Label target) {
```
Comment (wmdietl) [inline dataflow/src/org/checkerframework/dataflow/cfg/CFGBuilder.java]: Add some documentation.

File: javacutil/src/org/checkerframework/javacutil/trees/TreeBuilder.java [hunk 14]
```diff
@@ -63,6 +63,18 @@ public TreeBuilder(ProcessingEnvironment env) {
     }
 
     /**
+     * Copy an AST Tree.
+     *
+     * @param target a tree to be copied
+     * @return a copied tree
+     */
+    @SuppressWarnings("unchecked")
+    public <T extends Tree> T copy(T target) {
```
Comment (wmdietl) [inline javacutil/src/org/checkerframework/javacutil/trees/TreeBuilder.java]: Why is the parameter called `target` and not `source` or `input`? What is it the target of?

File: javacutil/src/org/checkerframework/javacutil/trees/TreeBuilder.java [hunk 15]
```diff
@@ -63,6 +63,18 @@ public TreeBuilder(ProcessingEnvironment env) {
     }
 
     /**
+     * Copy an AST Tree.
```
Comment (wmdietl) [inline javacutil/src/org/checkerframework/javacutil/trees/TreeBuilder.java]: Explain that it is a copy with all types and symbols preserved.

**Comment by wmdietl** [comment]
@bitterfox Can you please address the remaining comments and merge upstream to resolve the conflicts?

**Comment by wmdietl** [comment]
@bitterfox Will you have a chance to address these comments?

**Comment by bitterfox** [comment]
Yes! I'm now working on it and mostly done.
I can update this soon.

**Comment by bitterfox** [comment]
Hi, @wmdietl 
I've now fixed. Please check this again.
I'm now working on another PR and also fix it soon.

File: checker/tests/nullness/flow/Issue341.java [hunk 16]
```diff
@@ -1,18 +1,12 @@
 // Test case for issue #341:
 // https://github.com/typetools/checker-framework/issues/341
 
-// @skip-test
```
[the remark this answers is not in the document]
Comment (wmdietl) [reply checker/tests/nullness/flow/Issue341.java]: Can you add a comment to #341 what work remains to do for it?

**Pull request merged**